[2D][TP] Add DDP TP integration and unit test

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103604

Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

Commit:
Add DDP TP integration test

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46731700).

Commit:
Update on "Add DDP TP integration test"


Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1166,9 +1166,26 @@ def __setstate__(self, state):
+        from torch.distributed._tensor import DTensor
+        import functools
+        def _dtensor_ext(param_name, param):
+            def tensor_hook(param, grad):
```
what's the ordering of these hooks with respect to the Reducer installed hooks? How do we guarantee this ordering?

fduwjj replied:
This is not needed, actually. I haven't cleaned it up. I need to polish this PR a little bit more.

Commit:
Update on "[2D][TP] Add DDP TP integration and unit test"


Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

[ghstack-poisoned]

fduwjj committed:
Update on "[2D][TP] Add DDP TP integration and unit test"


Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

[ghstack-poisoned]

fduwjj committed:
Update on "[2D][TP] Add DDP TP integration and unit test"


Differential Revision: [D46731700](https://our.internmc.facebook.com/intern/diff/D46731700)

[ghstack-poisoned]

fduwjj commented on test/distributed/tensor/parallel/test_tp_ddp_2d.py:
```diff
@@ -0,0 +1,173 @@
+
+    twod_model = DDP(twod_model, process_group=dp_pg, find_unused_parameters=True)
+
+    # We also need a pre-forward hook to ensure gradients get propagated back.
```
yes. it's called every forward.

fduwjj commented on test/distributed/tensor/parallel/test_tp_ddp_2d.py:
```diff
@@ -0,0 +1,173 @@
+
+    # We need to find a way for DDP to replicate local tensor of DTensor.
+    for name, param in twod_model.named_parameters():
+        if isinstance(param, DTensor):
```
yes. Not need.

fduwjj commented:
Cannot ghstack import so created a new polished one in https://github.com/pytorch/pytorch/pull/106583.
